tests/functional/complete_sign_in.js

```diff
@@ -0,0 +1,136 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define([
+  'intern',
+  'intern!object',
+  'require',
+  'app/scripts/lib/constants',
+  'tests/lib/restmail',
+  'tests/lib/helpers',
+  'tests/functional/lib/helpers',
+  'tests/functional/lib/fx-desktop'
+], function (intern, registerSuite, require, Constants, restmail, TestHelpers, FunctionalHelpers, FxDesktopHelpers) {
+  var config = intern.config;
+  var EMAIL_SERVER_ROOT = config.fxaEmailRoot;
+  var PAGE_COMPLETE_SIGNIN_URL = config.fxaContentRoot + 'complete_signin';
+  var PAGE_SIGNIN_URL = config.fxaContentRoot + 'signin?context=fx_desktop_v1&service=sync';
+  var PASSWORD = 'password';
+  var user;
+  var email;
+  var code;
+  var uid;
+
+  var thenify = FunctionalHelpers.thenify;
+
+  var clearBrowserState = thenify(FunctionalHelpers.clearBrowserState);
+  var createUser = FunctionalHelpers.createUser;
+  var fillOutSignIn = thenify(FunctionalHelpers.fillOutSignIn);
+  var listenForFxaCommands = FxDesktopHelpers.listenForFxaCommands;
+  var noSuchElement = thenify(FunctionalHelpers.noSuchElement);
+  var openPage = thenify(FunctionalHelpers.openPage);
+  var testElementExists = FunctionalHelpers.testElementExists;
+  var testIsBrowserNotified = thenify(FxDesktopHelpers.testIsBrowserNotifiedOfMessage);
+  var testIsBrowserNotifiedOfLogin = thenify(FxDesktopHelpers.testIsBrowserNotifiedOfLogin);
+
+  var createRandomHexString = TestHelpers.createRandomHexString;
+
+  var setupTest = thenify(function (context) {
+    return this.parent
+      .then(createUser(email, PASSWORD, {preVerified: true}))
+      .then(openPage(context, PAGE_SIGNIN_URL, '#fxa-signin-header'))
+      .execute(listenForFxaCommands)
+      .then(fillOutSignIn(context, email, PASSWORD))
+      .then(testIsBrowserNotified(context, 'can_link_account'))
+      .then(testIsBrowserNotifiedOfLogin(context, email, {checkVerified: false}))
+      .then(testElementExists('#fxa-confirm-signin-header'))
+      .then(restmail(EMAIL_SERVER_ROOT + '/mail/' + user))
+      .then(function (emails) {
+        code = emails[0].headers['x-verify-code'];
+        uid = emails[0].headers['x-uid'];
+      });
+  });
+
+
+  registerSuite({
+    name: 'complete_sign_in',
+
+    beforeEach: function () {
+      email = TestHelpers.createEmail('sync{id}');
+      user = TestHelpers.emailToUser(email);
+      return this.remote
+        .then(clearBrowserState(this))
+        .then(setupTest(this, true));
+    },
+
+    'open verification link with malformed code': function () {
+      code = createRandomHexString(Constants.CODE_LENGTH - 1);
+      var url = PAGE_COMPLETE_SIGNIN_URL + '?uid=' + uid + '&code=' + code;
+
+      return this.remote
+        .get(require.toUrl(url))
+
+        // a successful user is immediately redirected to the
+        // sign-in-complete page.
+        .findById('fxa-verification-link-damaged-header')
+        .then(noSuchElement(this, '#resend'))
```

Comment:
This sequence is subtly incorrect. What it's saying is "Ensure there is no #resend element inside of the #fxa-verification-link-damaged-header" element. Since the header shouldn't contain the resend button, this test is not going to fail even if a resend button is on the page.

Can you use [`testElementExists`](https://github.com/mozilla/fxa-content-server/blob/025ed9141a735a1395b9ae8a39f575b10a410171/tests/functional/lib/helpers.js#L1102) on `#fxa-verification-link-damaged-header`, then line 77 won't need to change, and line 78 can go away.